pkg/controller/acmechallenges/sync.go

```diff
@@ -267,7 +268,13 @@ func (c *controller) handleFinalizer(ctx context.Context, ch *cmacme.Challenge)
 	}()
 
 	if !ch.Status.Processing {
-		return nil
+		if ch.Status.State == "expired" {
+			// When a challenge is "expired", no more work can be done and it should be cleaned up and deleted
+			expiredDeleteTimestamp := metav1.Now()
+			ch.SetDeletionTimestamp(&expiredDeleteTimestamp)
+		} else {
+			return nil
+		}
```

Comment:
nitpick: Usually in Go we'd write this slightly differently to return as early as possible and avoid nested code. [This website](https://100go.co/#unnecessary-nested-code-2) explains it well!

I've not looked into the implementation of this or whether the logic makes sense - just reviewing the Go since that's what you asked [on slack](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1724940962669879)! 😁 

```suggestion
		if ch.Status.State != "expired" {
			return nil
		}
		
		// When a challenge is "expired", no more work can be done and it should be cleaned up and deleted
		expiredDeleteTimestamp := metav1.Now()
		ch.SetDeletionTimestamp(&expiredDeleteTimestamp)
```

That said, comparing this code to the previous, I suspect we want to always return nil (even if we set the deletion timestamp)? Which would lead to the code below:

```suggestion
		if ch.Status.State == "expired" {
			// When a challenge is "expired", no more work can be done and it should be cleaned up and deleted
			expiredDeleteTimestamp := metav1.Now()
			ch.SetDeletionTimestamp(&expiredDeleteTimestamp)
		}

		return nil
```

(which is very similar to what you originall wrote!)

Reply:
I believe this will return from the function before the call to `CleanUp`, which is where this value would be resolved.

Reply:
Fair enough! In that case I think the first suggestion block would be appropriate? Totally possible I'm missing something though!